### ISSUE TITLE
ReferenceFile .validate is causing some errors in pipeline processing

### ISSUE THREAD

**Issue by hbushouse** [STScI-JWST/jwst#2139]
The changes in #2130 included changing all of the reference file data models so that they now sub-class ReferenceFileModel. In turn, ReferenceFileModel contains a .validate() method that checks to see if all of the reference file specific keywords, like AUTHOR, PEDIGREE, DESCRIP, etc., are contained in the model and throws an AssertionError if any of them are missing. This only happens when a ref file model is created.

But this has already been seen to cause problems in pipeline processing when a reference file is used that's missing one of those keywords. The initial loading and use of the reference file is OK, because validation only happens when a model is created, but there are at least two steps in the pipeline that create new reference file models from scratch: `dark_current` and `flat_field`. And when those steps create the new ref file model they also run .update() on the new model to copy over meta data from the original input ref file model. So if the original input ref file was missing one of those keywords, the validation that occurs on the new model throws an error.

None of these keywords is needed for processing within the cal pipeline and hence it should not cause an error when they aren't present. At the very least the AssertionError should be downgraded to a simple warning.

**Comment by jaytmiller** [comment]
Starting a recap of our e-mail discussion,  excerpting the solution oriented comments:

**Comment by jaytmiller** [comment]
Details from Howard:

1) It's all due to a relatively small number of NIRCam ref files:

  - jwst_nircam_dark_0031.fits is missing the AUTHOR keyword
  - jwst_nircam_flat_0002, 5, 6, 7, 8, 9.fits are missing the DESCRIP keyword

**Comment by jaytmiller** [comment]
Bernie:

If it was my decision, I'd catch the assertion error and convert it to a warning.

**Comment by jaytmiller** [comment]
Nadia:

FWIW, the ReferenceModel.validate() method was written as a helper method for INS to run afterthey have created their model. It was unfortunately named "validate" and overwrites the Datamodel.validate method.

We can rename that method to something else and this will fix the problem.

The unfortunate naming of this method is also a problem because it overwrites the original DataModel.validate and so the schemas are not validated (as Todd noticed some time ago). There's a github issue for this to be changed but I haven't worked on it yet.

This only showed up now because reference file models now inherit from ReferenceFileModel (which they shuold have been doing to start with) and not DataModel.


So, in short, I think ReferenceFileModel.validate should be renamed.

**Comment by jaytmiller** [comment]
Todd:  Warnings may be the sweet spot since that will still help INS create good files and will permit older non-compliant files to be used.   (crossing emails with Nadia who agreed.)

**Comment by nden** [comment]
#1965 is related

**Comment by jaytmiller** [comment]
Todd:  Fatal errors trying to propagate invalid values during calibration has been an unforeseen problem with our data models validation policies.   There are really two use cases:

1.  **Calibrators** should not use invalid values,  but may need to propagate files with invalid values or create dynamically modified versions.

2.  **Reference creators** should not create files with invalid or missing values (without knowing it) since these will cause repeat warnings or errors every time they're used.

These cases are in conflict;   issuing warnings is one compromise which works for these assertions but may not (didn't?) work for other models violations which occurred in the past.

Fixing or dropping old reference files is a possibility,  but the issue of backward compatibility will recur whenever models are retroactively made stricter.

Fully modeling references provides a format independent language (meta.x.y = ) for creating them with additional checks providing added value.

**Comment by jaytmiller** [comment]
One aspect of these .validate() functions is that they are not schema based,  they are ad hoc code based.  I think advanced checks are great,  don't get me wrong,  so that isn't bad.   But...  we may want to think about a unified way of handling validation errors,  so that things like PASS_INVALID and STRICT work in intuitive ways for all kinds of validate() errors,  not just schema violations.

**Comment by hbushouse** [comment]
I agree. We don't want validation implemented in two different places or ways. If something changes (new, deleted, renamed keyword), then we have to make changes in 2 places and that's a recipe for trouble. We need to finally figure out a way to mark keywords as required in the data model schemas and let all of the validation take place there. That also gives the user some control, via the setting of PASS_INVALID and STRICT. Having the checks hardwired in a validate function does not give you that capability.

**Comment by nden** [comment]
@jaytmiller I agree. I am looking into ways to do this but first I need to solve a problem with the wcs reference files schemas.
